concepts/iterators/introduction.md

```diff
@@ -0,0 +1,173 @@
+while let Some(people) = peoples.next() {
+    println!("{}", people);
+}
+```
+
+## `iter` and `iter_mut` methods
+
+Most Rust collection types (such as `Vec`, `HashMap`, `HashSet`) provide `iter` and `iter_mut` methods.
+For an iterator that produces values of type `T`, these two methods perform the following by convention:
+
+- The iterator returned by `iter` will yield immutably borrowed `&T`.
+- The iterator returned by `iter_mut` will yield mutably borrowed `&mut T`.
+
+```rust
+let mut v = vec![1, 2, 3];
+
+// Using `iter` method
+let mut iterator = v.iter();
+assert_eq!(iterator.next(), Some(&1));
+assert_eq!(iterator.next(), Some(&2));
+assert_eq!(iterator.next(), Some(&3));
+assert_eq!(iterator.next(), None);
+
+// Using `iter_mut` method
+let mut muterator = v.iter_mut();
+assert_eq!(muterator.next(), Some(&mut 1));
+assert_eq!(muterator.next(), Some(&mut 2));
+assert_eq!(muterator.next(), Some(&mut 3));
+assert_eq!(muterator.next(), None);
+```
+
+## Difference between `iter`, `iter_mut`, and `into_iter`
+
+Now that we've covered `iter`, `iter_mut`, and `into_iter`, what are the differences?
+Let's amend our previous bullet points:
+
+- The iterator returned by `iter` will yield immutably borrowed `&T`.
+- The iterator returned by `iter_mut` will yield mutably borrowed `&mut T`.
+- The iterator returned by `into_iter` may yield `T`, `&T` or `&mut T`, depending on its context.
+
+Wait a second, what do we mean by _depending on its context_?
+Just what `into_iter` is invoked on.
+Let's clarify using the following bullet points:
+
+- `(&v).into_iter()` is equal to `v.iter()` => returns `&T`
+- `(&mut v).into_iter()` is equal to `v.iter_mut()` => returns `&mut T`
+- `v.into_iter()` => returns `T` and moves the values
+
+This is most clear in for loops:
+
+- `for num in &v {...}` => returns `&T`
+- `for num in &mut v {...}` => returns `&mut T`
+- `for num in v {...}` => returns `T` and moves the values
+
+Unless we are working with for loops, it's always clearer to write `v.iter()` instead of `(&v).into_iter()`.
+Therefore, there are ergonomic reasons for preferring `iter` and `iter_mut` over `into_iter`.
+
+## Using built-in methods for iterators
+
+Iterators provide numerous built-in methods, such as `map`, `filter`, `collect`.
+These methods are sometimes used with closures.
+
+Using `map` with a closure to transform a vector into a different one:
+
+```rust
+let v = [1, 2, 3];
+let v2 iter = v.iter().map(|x| 2 * x);
+println!("{:?}", v2);
+// => [2, 4, 6]
+```
+
+Using `filter` with a closure to filter for certain attributes and then use `collect` to return a collection again:
+
+```rust
+let lor = vec!["Ainur", "Elves", "Men", "Dwarves", "Hobbits", "Orcs", "Dragons"];
+let short_people = ["Dwarves", "Hobbits"];
+
+let short_filter: Vec<&str> = lor
+    .into_iter()
+    .filter(|people| short_people.contains(people))
+    .collect();
+
+println!("{:?}", short_filter);
+// => ["Dwarves", "Hobbits"]
+```
```

Comment:
~~~suggestion
`filter` transforms the iterator into one which only returns those values for which the closure evaluates to `true`.
`count` consumes the iterator, returning the number of items produced.

```rust
let lor = vec!["Ainur", "Elves", "Men", "Dwarves", "Hobbits", "Orcs", "Dragons"];
let short_people = ["Dwarves", "Hobbits"];
let num_tall_people = lor.iter().filter(|people| !short_people.contains(people)).count();
assert_eq!(num_tall_people, 5);
```

`flat_map` transforms an iterator by transforming each item into a new iterator via the provided closure, then flattening all the produced new iterators by chaining them together.
`collect` produces a new collection, as long as the collection type implements [`FromIterator`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html).

```rust
let words = ["alpha", "beta", "gamma"];

// chars() returns an iterator
let merged: String = words.iter()
                          .flat_map(|s| s.chars())
                          .collect();
assert_eq!(merged, "alphabetagamma");
```

The [`Iterator` trait](https://doc.rust-lang.org/std/iter/trait.Iterator.html) provides _many_ such useful iterator transforms; it's worth reading the documentation and familiarizing yourself with them.
~~~

It's worth talking about a few other examples of iterator methods IMO, to pique interest.